### ISSUE TITLE
Enhancement: Compatibility with Pandas DataFrame?

### ISSUE THREAD

**Issue by bscully** [all-umass/metric-learn#77]
For an enhancement, are there plans to also accept pandas dataframes as inputs?
This project is something I could help with.

**Comment by perimosocordiae** [comment]
Dataframes are a bit tricky, because they kind-of act like normal arrays, but not always.

That said, I'm happy to take a look at PRs that add support for them!

**Comment by terrytangyuan** [comment]
Perhaps we could just add a layer to convert pandas data types to format that metric-learn can currently take, similar to what we did in XGBoost [here](https://github.com/dmlc/xgboost/blob/96826a35155d42f78cd4253b3d134a208dbbaccb/python-package/xgboost/core.py#L201).

**Comment by wdevazelhes** [comment]
For the next release, we thought of supporting pandas dataframes as input through the use of scikit-learn's `check_array` function (http://scikit-learn.org/stable/modules/generated/sklearn.utils.check_array.html), but it does not keep the feature names in memory (scikit-learn does not either for now, except for some particular transformers IIRC), which we maybe would need to provide feature-wise logs (like feature importances etc) 
Looking at XGBoost it's interesting to see that the feature names are memorized in the model, which allows to directly plot feature importances with the right names using `plot_importance`
